programs/psa/psa_hash.c

```diff
@@ -0,0 +1,172 @@
+/*
+ *  Example computing a SHA-256 hash using the PSA Crypto API
+ *
+ *  The example computes the SHA-256 hash of a test string using the
+ *  one-shot API call psa_hash_compute() and the using multi-part
+ *  operation, which requires psa_hash_setup(), psa_hash_update() and
+ *  psa_hash_finish(). The multi-part operation is popular on embedded
+ *  devices where a rolling hash needs to be computed.
+ *
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+#include "psa/crypto.h"
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "mbedtls/build_info.h"
+#include "mbedtls/platform.h"
+
+/* Information about hashing with the PSA API can be
+ * found here:
+ * https://arm-software.github.io/psa-api/crypto/1.1/api/ops/hashes.html
+ *
+ * The algorithm used by this demo is SHA 256.
+ * Please see include/psa/crypto_values.h to see the other
+ * algorithms that are supported by Mbed TLS.
+ * If you switch to a different algorithm you will need to update
+ * the hash data in the SAMPLE_HASH_DATA macro below. */
+
+#define HASH_ALG PSA_ALG_SHA_256
+
+#define SAMPLE_HASH_DATA {                                                      \
+        0x7f, 0x83, 0xb1, 0x65, 0x7f, 0xf1, 0xfc, 0x53, 0xb9, 0x2d, 0xc1, 0x81, \
+        0x48, 0xa1, 0xd6, 0x5d, 0xfc, 0x2d, 0x4b, 0x1f, 0xa3, 0xd6, 0x77, 0x28, \
+        0x4a, 0xdd, 0xd2, 0x00, 0x12, 0x6d, 0x90, 0x69 \
+}
+
+const uint8_t sample_hash[] = SAMPLE_HASH_DATA;
+const size_t sample_hash_len = sizeof(sample_hash);
+
+const uint8_t sample_message[] = "Hello World!";
+/* sample_message is terminated with a null byte which is not part of
+ * the message itself so we make sure to subtract it in order to get
+ * the message length. */
+const size_t sample_message_length = sizeof(sample_message) - 1;
```

Comment:
I think the sample program would read better if this sample message and its length appeared before the expected hash value. And note "expected" - again, it would read better with `EXPECTED_HASH_DATA`, `expected_hash`, and `expected_hash_len`.

Also, these variables are outside the macro guards, which means if it's compiled without `MBEDTLS_PSA_CRYPTO_C` or `PSA_WANT_ALG_SHA_256`, some compilers will give warnings about unused variables - at least, they should, but now I see that they're not `static`, which they probably should be - I don't see a reason to have them global. So they should be `static` and within the compilation guard.

Finally, I wonder if it's worth having the "interesting" version of `main()` - i.e. the one with all the good stuff in it - before the un-interesting "these aren't defined" version. That would mean the guards become `#if defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_ALG_SHA_256)` and therefore match the `#endif` at the bottom of the file, which isn't currently right.

However, that is the opposite of the way the example programs are currently done, so I'd be keen to see what other reviewers think.

That is, how does the following layout seem?

```
:
#include "mbedtls/build_info.h"
#include "mbedtls/platform.h"

#if defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_ALG_SHA_256)

/* Information about hashing with the PSA API can be found here:
 * https://arm-software.github.io/psa-api/crypto/1.1/api/ops/hashes.html
 :
 * the hash output in the EXPECTED_HASH_DATA macro below. */

#define HASH_ALG PSA_ALG_SHA_256

const uint8_t sample_message[] = "Hello World!";
/* sample_message is terminated with a null byte which is not part of
 * the message itself so we make sure to subtract it in order to get
 * the message length. */
const size_t sample_message_length = sizeof(sample_message) - 1;

#define EXPECTED_HASH_DATA {                                                \
    0x7f, 0x83, 0xb1, 0x65, 0x7f, 0xf1, 0xfc, 0x53, 0xb9, 0x2d, 0xc1, 0x81, \
    0x48, 0xa1, 0xd6, 0x5d, 0xfc, 0x2d, 0x4b, 0x1f, 0xa3, 0xd6, 0x77, 0x28, \
    0x4a, 0xdd, 0xd2, 0x00, 0x12, 0x6d, 0x90, 0x69 \
}

const uint8_t expected_hash[] = EXPECTED_HASH_DATA;
const size_t expected_hash_len = sizeof(expected_hash);

int main(void)
{
    psa_status_t status;
    uint8_t hash[PSA_HASH_LENGTH(HASH_ALG)];
    :
    :
    return EXIT_FAILURE;
}

#else   /* MBEDTLS_PSA_CRYPTO_C && PSA_WANT_ALG_SHA_256 */

int main(void)
{
    mbedtls_printf("MBEDTLS_PSA_CRYPTO_C and PSA_WANT_ALG_SHA_256"
                   "not defined.\r\n");
    return EXIT_SUCCESS;
}

#endif  /* MBEDTLS_PSA_CRYPTO_C && PSA_WANT_ALG_SHA_256 */
```

Reply:
I think I have implemented a compromise. PTAL when you have the bandwidth